In [1]:
import Acqu as aq

# Acqu Module
The Acqu module parses in Mk1 and Mk2 data files which can be read event by event. A numpy array is made of form, (ADC,Value) for each event stored as **adcArray**. If the event contains scaler or epics data this can also be read but hasn't been tested recently.


In [2]:
#inFile = '/w/work0/mainz/2008-09_Recoil/CB_19156.dat'
#inFile = '/w/work0/mainz/2018-08_Timepix3-Acqu/Timepix_49.dat'
inFile = '/w/work1/home/simong/TP3_15.dat'
#inFile = '/scratch/CBTaggTAPS_16271.dat'
aq.openFile(inFile)

Mk2 Data


# Define functions
The module allows any arbitrary function to be run on each event. This is passed as an argument to the **runFunction** function in the Acqu module. The other arguments are the number of the first and last event to be read.

Here the function just prints the event number and the raw adcArray:


In [3]:
def printRAW():
    print('number',aq.eventNo)
    print(aq.adcArray)

In [4]:
aq.runFunction(printRAW)#,0,20000)

('number', 1)
[(200,  2896) (201, 13686) (202,    12) (203,     0)]
('number', 2)
[(200, 25731) (201, 13688) (202,    12) (203,     0)]
('number', 3)
[(200, 20200) (201, 13690) (202,    10) (203,     0)]
('number', 4)
[(200, 27646) (201, 13914) (202,  1198) (203,     0)]
('number', 5)
[(200, 54118) (201, 13919) (202,    29) (203,     0)]
('number', 6)
[(200, 30795) (201, 13921) (202,     9) (203,     0)]
('number', 7)
[(200, 44845) (201, 13923) (202,    12) (203,     0)]
('number', 8)
[(200, 65326) (201, 13928) (202,    28) (203,     0)]
('number', 9)
[(200, 16170) (201, 13929) (202,     1) (203,     0)]
('number', 10)
[(200, 42813) (201, 13931) (202,    13) (203,     0)]
('number', 11)
[(200, 44690) (201, 13932) (202,     6) (203,     0)]
('number', 12)
[(200, 56117) (201, 13933) (202,     6) (203,     0)]
('number', 13)
[(200,   416) (201, 13936) (202,    11) (203,     0)]
('number', 14)
[(200, 19677) (201, 13939) (202,    18) (203,     0)]
('number', 15)
[(200, 61515) (201, 13943) (

[(200, 43300) (201, 14839) (202,    14) (203,     0)]
('number', 293)
[(200, 26771) (201, 14840) (202,     4) (203,     0)]
('number', 294)
[(200, 29623) (201, 14845) (202,    27) (203,     0)]
('number', 295)
[(200, 12364) (201, 14847) (202,     9) (203,     0)]
('number', 296)
[(200, 61690) (201, 14849) (202,    15) (203,     0)]
('number', 297)
[(200, 57839) (201, 14851) (202,    10) (203,     0)]
('number', 298)
[(200, 41875) (201, 14852) (202,     4) (203,     0)]
('number', 299)
[(200, 47256) (201, 14855) (202,    17) (203,     0)]
('number', 300)
[(200, 20827) (201, 14857) (202,     8) (203,     0)]
('number', 301)
[(200, 52079) (201, 14857) (202,     3) (203,     0)]
('number', 302)
[(200, 15188) (201, 14858) (202,     2) (203,     0)]
('number', 303)
[(200, 63649) (201, 14858) (202,     4) (203,     0)]
('number', 304)
[(200, 34095) (201, 14862) (202,    19) (203,     0)]
('number', 305)
[(200, 41468) (201, 14863) (202,     6) (203,     0)]
('number', 306)
[(200, 20441) (201, 

('number', 563)
[(200, 39693) (201, 15647) (202,     1) (203,     0)]
('number', 564)
[(200, 45741) (201, 15651) (202,    22) (203,     0)]
('number', 565)
[(200, 61152) (201, 15654) (202,    17) (203,     0)]
('number', 566)
[(200, 59520) (201, 15660) (202,    32) (203,     0)]
('number', 567)
[(200, 44880) (201, 15672) (202,    63) (203,     0)]
('number', 568)
[(200, 25195) (201, 15674) (202,     9) (203,     0)]
('number', 569)
[(200, 57279) (201, 15674) (202,     3) (203,     0)]
('number', 570)
[(200, 11040) (201, 15677) (202,    12) (203,     0)]
('number', 571)
[(200, 57114) (201, 15677) (202,     4) (203,     0)]
('number', 572)
[(200, 38661) (201, 15685) (202,    41) (203,     0)]
('number', 573)
[(200, 17377) (201, 15687) (202,     9) (203,     0)]
('number', 574)
[(200, 14251) (201, 15688) (202,     5) (203,     0)]
('number', 575)
[(200,   746) (201, 15690) (202,    10) (203,     0)]
('number', 576)
[(200,  3769) (201, 15691) (202,     5) (203,     0)]
('number', 577)
[(20

('number', 845)
[(200, 51760) (201, 16625) (202,    15) (203,     0)]
('number', 846)
[(200,  3287) (201, 16626) (202,     2) (203,     0)]
('number', 847)
[(200, 64848) (201, 16626) (202,     5) (203,     0)]
('number', 848)
[(200, 42004) (201, 16629) (202,    14) (203,     0)]
('number', 849)
[(200,  6665) (201, 16630) (202,     2) (203,     0)]
('number', 850)
[(200, 56111) (201, 16630) (202,     4) (203,     0)]
('number', 851)
[(200, 47740) (201, 16635) (202,    26) (203,     0)]
('number', 852)
[(200, 11297) (201, 16636) (202,     3) (203,     0)]
('number', 853)
[(200, 52758) (201, 16636) (202,     3) (203,     0)]
('number', 854)
[(200,  6514) (201, 16639) (202,    12) (203,     0)]
('number', 855)
[(200, 13711) (201, 16642) (202,    17) (203,     0)]
('number', 856)
[(200, 28381) (201, 16646) (202,    23) (203,     0)]
('number', 857)
[(200, 36365) (201, 16649) (202,    16) (203,     0)]
('number', 858)
[(200,  8525) (201, 16651) (202,     9) (203,     0)]
('number', 859)
[(20

('number', 1135)
[(200, 42539) (201, 17497) (202,    11) (203,     0)]
('number', 1136)
[(200, 22683) (201, 17498) (202,     4) (203,     0)]
('number', 1137)
[(200, 38659) (201, 17498) (202,     1) (203,     0)]
('number', 1138)
[(200, 50488) (201, 17500) (202,    12) (203,     0)]
('number', 1139)
[(200, 52964) (201, 17511) (202,    59) (203,     0)]
('number', 1140)
[(200, 65086) (201, 17516) (202,    28) (203,     0)]
('number', 1141)
[(200, 24309) (201, 17517) (202,     2) (203,     0)]
('number', 1142)
[(200, 22271) (201, 17521) (202,    21) (203,     0)]
('number', 1143)
[(200, 44929) (201, 17526) (202,    29) (203,     0)]
('number', 1144)
[(200,  3432) (201, 17533) (202,    34) (203,     0)]
('number', 1145)
[(200, 41847) (201, 17535) (202,    13) (203,     0)]
('number', 1146)
[(200, 45273) (201, 17546) (202,    59) (203,     0)]
('number', 1147)
[(200, 31485) (201, 17552) (202,    31) (203,     0)]
('number', 1148)
[(200, 14682) (201, 17559) (202,    36) (203,     0)]
('numb

[(200,  8510) (201, 18395) (202,    17) (203,     0)]
('number', 1409)
[(200, 10829) (201, 18397) (202,    11) (203,     0)]
('number', 1410)
[(200, 29433) (201, 18398) (202,     7) (203,     0)]
('number', 1411)
[(200, 36526) (201, 18400) (202,    11) (203,     0)]
('number', 1412)
[(200, 45928) (201, 18401) (202,     6) (203,     0)]
('number', 1413)
[(200, 12918) (201, 18403) (202,     8) (203,     0)]
('number', 1414)
[(200, 33764) (201, 18405) (202,    13) (203,     0)]
('number', 1415)
[(200, 58401) (201, 18406) (202,     7) (203,     0)]
('number', 1416)
[(200, 31725) (201, 18409) (202,    14) (203,     0)]
('number', 1417)
[(200, 53448) (201, 18415) (202,    34) (203,     0)]
('number', 1418)
[(200, 64480) (201, 18416) (202,     6) (203,     0)]
('number', 1419)
[(200,  9259) (201, 18419) (202,    12) (203,     0)]
('number', 1420)
[(200,  2922) (201, 18423) (202,    20) (203,     0)]
('number', 1421)
[(200, 58401) (201, 18423) (202,     5) (203,     0)]
('number', 1422)
[(200,

[(200, 38609) (201, 19123) (202,     5) (203,     0)]
('number', 1606)
[(200, 37274) (201, 19124) (202,     5) (203,     0)]
('number', 1607)
[(200, 34390) (201, 19125) (202,     5) (203,     0)]
('number', 1608)
[(200, 15877) (201, 19126) (202,     4) (203,     0)]
('number', 1609)
[(200, 29795) (201, 19127) (202,     6) (203,     0)]
('number', 1610)
[(200, 36904) (201, 19128) (202,     6) (203,     0)]
('number', 1611)
[(200, 59383) (201, 19137) (202,    50) (203,     0)]
('number', 1612)
[(200, 47035) (201, 19138) (202,     5) (203,     0)]
('number', 1613)
[(200, 13748) (201, 19140) (202,     8) (203,     0)]
('number', 1614)
[(200,  3421) (201, 19145) (202,    26) (203,     0)]
('number', 1615)
[(200, 17528) (201, 19148) (202,    17) (203,     0)]
('number', 1616)
[(200, 28284) (201, 19151) (202,    17) (203,     0)]
('number', 1617)
[(200, 50841) (201, 19153) (202,    12) (203,     0)]
('number', 1618)
[(200,  3667) (201, 19154) (202,     2) (203,     0)]
('number', 1619)
[(200,

('number', 1868)
[(200, 25782) (201, 19995) (202,     6) (203,     0)]
('number', 1869)
[(200, 41899) (201, 19996) (202,     7) (203,     0)]
('number', 1870)
[(200,  3801) (201, 20002) (202,    29) (203,     0)]
('number', 1871)
[(200, 11496) (201, 20005) (202,    17) (203,     0)]
('number', 1872)
[(200, 62071) (201, 20016) (202,    62) (203,     0)]
('number', 1873)
[(200, 20206) (201, 20018) (202,     8) (203,     0)]
('number', 1874)
[(200,  3213) (201, 20023) (202,    25) (203,     0)]
('number', 1875)
[(200, 32138) (201, 20023) (202,     2) (203,     0)]
('number', 1876)
[(200, 24609) (201, 20029) (202,    32) (203,     0)]
('number', 1877)
[(200, 47944) (201, 20033) (202,    23) (203,     0)]
('number', 1878)
[(200, 16509) (201, 20040) (202,    35) (203,     0)]
('number', 1879)
[(200, 31857) (201, 20041) (202,     7) (203,     0)]
('number', 1880)
[(200, 13793) (201, 20044) (202,    14) (203,     0)]
('number', 1881)
[(200, 61154) (201, 20047) (202,    20) (203,     0)]
('numb

('number', 2152)
[(200, 18383) (201, 20942) (202,    18) (203,     0)]
('number', 2153)
[(200, 44135) (201, 20942) (202,     2) (203,     0)]
('number', 2154)
[(200, 13994) (201, 20943) (202,     3) (203,     0)]
('number', 2155)
[(200, 52394) (201, 20945) (202,    14) (203,     0)]
('number', 2156)
[(200, 51979) (201, 20953) (202,    43) (203,     0)]
('number', 2157)
[(200,  3636) (201, 20954) (202,     1) (203,     0)]
('number', 2158)
[(200, 13914) (201, 20955) (202,     6) (203,     0)]
('number', 2159)
[(200, 54671) (201, 20956) (202,     9) (203,     0)]
('number', 2160)
[(200, 12756) (201, 20962) (202,    28) (203,     0)]
('number', 2161)
[(200, 39709) (201, 20963) (202,     8) (203,     0)]
('number', 2162)
[(200, 37259) (201, 20964) (202,     5) (203,     0)]
('number', 2163)
[(200, 10195) (201, 20965) (202,     3) (203,     0)]
('number', 2164)
[(200, 36710) (201, 20965) (202,     2) (203,     0)]
('number', 2165)
[(200, 36712) (201, 20969) (202,    22) (203,     0)]
('numb

('number', 2423)
[(200, 45839) (201, 21823) (202,     5) (203,     0)]
('number', 2424)
[(200, 10778) (201, 21825) (202,     8) (203,     0)]
('number', 2425)
[(200, 59786) (201, 21825) (202,     4) (203,     0)]
('number', 2426)
[(200, 30755) (201, 21830) (202,    24) (203,     0)]
('number', 2427)
[(200, 12764) (201, 21833) (202,    15) (203,     0)]
('number', 2428)
[(200, 53278) (201, 21834) (202,     8) (203,     0)]
('number', 2429)
[(200, 42476) (201, 21839) (202,    26) (203,     0)]
('number', 2430)
[(200, 36987) (201, 21843) (202,    21) (203,     0)]
('number', 2431)
[(200, 59197) (201, 21845) (202,    13) (203,     0)]
('number', 2432)
[(200, 22924) (201, 21846) (202,     2) (203,     0)]
('number', 2433)
[(200, 40981) (201, 21846) (202,     1) (203,     0)]
('number', 2434)
[(200, 26643) (201, 21850) (202,    21) (203,     0)]
('number', 2435)
[(200, 53694) (201, 21854) (202,    23) (203,     0)]
('number', 2436)
[(200, 48064) (201, 21855) (202,     5) (203,     0)]
('numb

('number', 2712)
[(200, 53494) (201, 22692) (202,     5) (203,     0)]
('number', 2713)
[(200,  2259) (201, 22696) (202,    18) (203,     0)]
('number', 2714)
[(200, 64840) (201, 22698) (202,    15) (203,     0)]
('number', 2715)
[(200, 23200) (201, 22700) (202,     8) (203,     0)]
('number', 2716)
[(200, 38048) (201, 22706) (202,    33) (203,     0)]
('number', 2717)
[(200, 58179) (201, 22708) (202,    12) (203,     0)]
('number', 2718)
[(200, 40611) (201, 22712) (202,    20) (203,     0)]
('number', 2719)
[(200, 48791) (201, 22714) (202,    12) (203,     0)]
('number', 2720)
[(200, 62590) (201, 22716) (202,    11) (203,     0)]
('number', 2721)
[(200, 37009) (201, 22729) (202,    68) (203,     0)]
('number', 2722)
[(200, 50842) (201, 22735) (202,    33) (203,     0)]
('number', 2723)
[(200, 43080) (201, 22738) (202,    15) (203,     0)]
('number', 2724)
[(200, 38642) (201, 22742) (202,    21) (203,     0)]
('number', 2725)
[(200,   636) (201, 22743) (202,     3) (203,     0)]
('numb

('number', 2976)
[(200, 54945) (201, 23614) (202,     3) (203,     0)]
('number', 2977)
[(200, 13665) (201, 23616) (202,     7) (203,     0)]
('number', 2978)
[(200, 49335) (201, 23621) (202,    29) (203,     0)]
('number', 2979)
[(200, 21165) (201, 23624) (202,    14) (203,     0)]
('number', 2980)
[(200,  9508) (201, 23631) (202,    37) (203,     0)]
('number', 2981)
[(200, 23850) (201, 23632) (202,     6) (203,     0)]
('number', 2982)
[(200, 34747) (201, 23642) (202,    55) (203,     0)]
('number', 2983)
[(200, 64707) (201, 23644) (202,    13) (203,     0)]
('number', 2984)
[(200, 12185) (201, 23647) (202,    11) (203,     0)]
('number', 2985)
[(200, 31601) (201, 23659) (202,    66) (203,     0)]
('number', 2986)
[(200, 27247) (201, 23661) (202,    10) (203,     0)]
('number', 2987)
[(200, 48156) (201, 23664) (202,    18) (203,     0)]
('number', 2988)
[(200, 27552) (201, 23668) (202,    20) (203,     0)]
('number', 2989)
[(200, 63223) (201, 23680) (202,    67) (203,     0)]
('numb

[(200,  7947) (201, 24444) (202,     2) (203,     0)]
('number', 3248)
[(200, 25377) (201, 24444) (202,     1) (203,     0)]
('number', 3249)
[(200, 17361) (201, 24448) (202,    21) (203,     0)]
('number', 3250)
[(200, 31137) (201, 24450) (202,    12) (203,     0)]
('number', 3251)
[(200, 33557) (201, 24451) (202,     5) (203,     0)]
('number', 3252)
[(200, 23620) (201, 24452) (202,     5) (203,     0)]
('number', 3253)
[(200, 35713) (201, 24453) (202,     6) (203,     0)]
('number', 3254)
[(200,  8088) (201, 24456) (202,    14) (203,     0)]
('number', 3255)
[(200, 43866) (201, 24457) (202,     8) (203,     0)]
('number', 3256)
[(200, 38255) (201, 24466) (202,    48) (203,     0)]
('number', 3257)
[(200, 40428) (201, 24471) (202,    27) (203,     0)]
('number', 3258)
[(200, 33136) (201, 24477) (202,    31) (203,     0)]
('number', 3259)
[(200, 59620) (201, 24479) (202,    13) (203,     0)]
('number', 3260)
[(200,  9088) (201, 24480) (202,     1) (203,     0)]
('number', 3261)
[(200,

('number', 3502)
[(200, 45074) (201, 25298) (202,    54) (203,     0)]
('number', 3503)
[(200, 59184) (201, 25300) (202,    12) (203,     0)]
('number', 3504)
[(200, 44713) (201, 25310) (202,    52) (203,     0)]
('number', 3505)
[(200, 15215) (201, 25316) (202,    30) (203,     0)]
('number', 3506)
[(200, 17705) (201, 25326) (202,    54) (203,     0)]
('number', 3507)
[(200, 49529) (201, 25326) (202,     2) (203,     0)]
('number', 3508)
[(200,   329) (201, 25331) (202,    23) (203,     0)]
('number', 3509)
[(200, 43772) (201, 25338) (202,    41) (203,     0)]
('number', 3510)
[(200, 26562) (201, 25340) (202,     9) (203,     0)]
('number', 3511)
[(200, 33571) (201, 25346) (202,    33) (203,     0)]
('number', 3512)
[(200, 21479) (201, 25348) (202,    10) (203,     0)]
('number', 3513)
[(200, 24052) (201, 25350) (202,    11) (203,     0)]
('number', 3514)
[(200, 65418) (201, 25351) (202,     8) (203,     0)]
('number', 3515)
[(200,  3059) (201, 25353) (202,     6) (203,     0)]
('numb

('number', 3759)
[(200, 59759) (201, 26136) (202,    24) (203,     0)]
('number', 3760)
[(200, 16300) (201, 26138) (202,     8) (203,     0)]
('number', 3761)
[(200, 26376) (201, 26142) (202,    22) (203,     0)]
('number', 3762)
[(200, 60781) (201, 26152) (202,    56) (203,     0)]
('number', 3763)
[(200, 37506) (201, 26163) (202,    57) (203,     0)]
('number', 3764)
[(200, 56759) (201, 26166) (202,    17) (203,     0)]
('number', 3765)
[(200, 36036) (201, 26168) (202,     9) (203,     0)]
('number', 3766)
[(200, 26159) (201, 26172) (202,    21) (203,     0)]
('number', 3767)
[(200, 62412) (201, 26176) (202,    24) (203,     0)]
('number', 3768)
[(200, 45571) (201, 26184) (202,    42) (203,     0)]
('number', 3769)
[(200, 26934) (201, 26185) (202,     4) (203,     0)]
('number', 3770)
[(200, 11892) (201, 26189) (202,    20) (203,     0)]
('number', 3771)
[(200, 35190) (201, 26190) (202,     7) (203,     0)]
('number', 3772)
[(200, 62449) (201, 26190) (202,     2) (203,     0)]
('numb

[(200, 33951) (201, 26683) (202,    28) (203,     0)]
('number', 3944)
[(200, 31312) (201, 26684) (202,     5) (203,     0)]
('number', 3945)
[(200, 51415) (201, 26685) (202,     7) (203,     0)]
('number', 3946)
[(200, 10983) (201, 26692) (202,    34) (203,     0)]
('number', 3947)
[(200, 21433) (201, 26695) (202,    17) (203,     0)]
('number', 3948)
[(200, 53919) (201, 26695) (202,     3) (203,     0)]
('number', 3949)
[(200,  1393) (201, 26699) (202,    17) (203,     0)]
('number', 3950)
[(200, 40867) (201, 26702) (202,    19) (203,     0)]
('number', 3951)
[(200, 25511) (201, 26703) (202,     4) (203,     0)]
('number', 3952)
[(200, 33422) (201, 26705) (202,    11) (203,     0)]
('number', 3953)
[(200, 53010) (201, 26705) (202,     2) (203,     0)]
('number', 3954)
[(200, 48235) (201, 26710) (202,    26) (203,     0)]
('number', 3955)
[(200, 47652) (201, 26712) (202,    11) (203,     0)]
('number', 3956)
[(200, 10768) (201, 26717) (202,    24) (203,     0)]
('number', 3957)
[(200,

('number', 4171)
[(200, 23349) (201, 27293) (202,     3) (203,     0)]
('number', 4172)
[(200, 12857) (201, 27302) (202,    47) (203,     0)]
('number', 4173)
[(200, 40503) (201, 27302) (202,     3) (203,     0)]
('number', 4174)
[(200, 28865) (201, 27304) (202,     9) (203,     0)]
('number', 4175)
[(200, 64817) (201, 27309) (202,    30) (203,     0)]
('number', 4176)
[(200, 39588) (201, 27312) (202,    14) (203,     0)]
('number', 4177)
[(200, 63708) (201, 27314) (202,    13) (203,     0)]
('number', 4178)
[(200, 50770) (201, 27315) (202,     4) (203,     0)]
('number', 4179)
[(200, 54562) (201, 27318) (202,    16) (203,     0)]
('number', 4180)
[(200, 57997) (201, 27320) (202,    11) (203,     0)]
('number', 4181)
[(200, 48721) (201, 27321) (202,     5) (203,     0)]
('number', 4182)
[(200, 29093) (201, 27322) (202,     4) (203,     0)]
('number', 4183)
[(200, 31549) (201, 27328) (202,    32) (203,     0)]
('number', 4184)
[(200, 62890) (201, 27329) (202,     8) (203,     0)]
('numb

('number', 4455)
[(200, 36165) (201, 28208) (202,    11) (203,     0)]
('number', 4456)
[(200, 54706) (201, 28208) (202,     1) (203,     0)]
('number', 4457)
[(200, 54919) (201, 28221) (202,    70) (203,     0)]
('number', 4458)
[(200, 48956) (201, 28223) (202,    10) (203,     0)]
('number', 4459)
[(200, 17474) (201, 28232) (202,    46) (203,     0)]
('number', 4460)
[(200, 59168) (201, 28236) (202,    24) (203,     0)]
('number', 4461)
[(200, 25942) (201, 28242) (202,    30) (203,     0)]
('number', 4462)
[(200, 24578) (201, 28248) (202,    32) (203,     0)]
('number', 4463)
[(200, 64131) (201, 28248) (202,     3) (203,     0)]
('number', 4464)
[(200, 56088) (201, 28250) (202,    10) (203,     0)]
('number', 4465)
[(200,  7070) (201, 28251) (202,     1) (203,     0)]
('number', 4466)
[(200, 27666) (201, 28251) (202,     2) (203,     0)]
('number', 4467)
[(200, 10881) (201, 28256) (202,    25) (203,     0)]
('number', 4468)
[(200, 56083) (201, 28259) (202,    20) (203,     0)]
('numb

('number', 4727)
[(200,  1107) (201, 29050) (202,    27) (203,     0)]
('number', 4728)
[(200, 27605) (201, 29052) (202,    13) (203,     0)]
('number', 4729)
[(200, 20735) (201, 29053) (202,     5) (203,     0)]
('number', 4730)
[(200, 50449) (201, 29054) (202,     8) (203,     0)]
('number', 4731)
[(200, 59002) (201, 29065) (202,    59) (203,     0)]
('number', 4732)
[(200,  6456) (201, 29071) (202,    28) (203,     0)]
('number', 4733)
[(200, 51729) (201, 29073) (202,    14) (203,     0)]
('number', 4734)
[(200, 16557) (201, 29076) (202,    14) (203,     0)]
('number', 4735)
[(200, 38638) (201, 29078) (202,    12) (203,     0)]
('number', 4736)
[(200, 33013) (201, 29081) (202,    16) (203,     0)]
('number', 4737)
[(200, 53976) (201, 29082) (202,     7) (203,     0)]
('number', 4738)
[(200, 57706) (201, 29083) (202,     5) (203,     0)]
('number', 4739)
[(200, 51586) (201, 29084) (202,     5) (203,     0)]
('number', 4740)
[(200, 41171) (201, 29088) (202,    21) (203,     0)]
('numb

('number', 4979)
[(200, 62391) (201, 29944) (202,    13) (203,     0)]
('number', 4980)
[(200, 22846) (201, 29947) (202,    13) (203,     0)]
('number', 4981)
[(200, 59303) (201, 29949) (202,    13) (203,     0)]
('number', 4982)
[(200, 35452) (201, 29955) (202,    30) (203,     0)]
('number', 4983)
[(200,  3503) (201, 29957) (202,     8) (203,     0)]
('number', 4984)
[(200, 10991) (201, 29959) (202,    12) (203,     0)]
('number', 4985)
[(200, 49363) (201, 29961) (202,    14) (203,     0)]
('number', 4986)
[(200, 44430) (201, 29963) (202,    10) (203,     0)]
('number', 4987)
[(200, 62025) (201, 29966) (202,    17) (203,     0)]
('number', 4988)
[(200, 20753) (201, 29979) (202,    66) (203,     0)]
('number', 4989)
[(200, 51428) (201, 29980) (202,     8) (203,     0)]
('number', 4990)
[(200, 15933) (201, 29991) (202,    56) (203,     0)]
('number', 4991)
[(200, 23392) (201, 29992) (202,     6) (203,     0)]
('number', 4992)
[(200,   253) (201, 29998) (202,    30) (203,     0)]
('numb

('number', 5245)
[(200, 21979) (201, 30794) (202,    13) (203,     0)]
('number', 5246)
[(200, 31806) (201, 30807) (202,    70) (203,     0)]
('number', 5247)
[(200, 61235) (201, 30808) (202,     8) (203,     0)]
('number', 5248)
[(200, 31038) (201, 30811) (202,    13) (203,     0)]
('number', 5249)
[(200, 11316) (201, 30817) (202,    31) (203,     0)]
('number', 5250)
[(200, 14206) (201, 30819) (202,    10) (203,     0)]
('number', 5251)
[(200, 65513) (201, 30823) (202,    26) (203,     0)]
('number', 5252)
[(200, 27558) (201, 30825) (202,     8) (203,     0)]
('number', 5253)
[(200,  7479) (201, 30833) (202,    41) (203,     0)]
('number', 5254)
[(200, 17537) (201, 30838) (202,    27) (203,     0)]
('number', 5255)
[(200, 34293) (201, 30839) (202,     7) (203,     0)]
('number', 5256)
[(200, 12971) (201, 30841) (202,     9) (203,     0)]
('number', 5257)
[(200, 39084) (201, 30846) (202,    29) (203,     0)]
('number', 5258)
[(200, 46164) (201, 30847) (202,     6) (203,     0)]
('numb

('number', 5537)
[(200, 39437) (201, 31677) (202,    20) (203,     0)]
('number', 5538)
[(200, 54610) (201, 31684) (202,    39) (203,     0)]
('number', 5539)
[(200, 64109) (201, 31688) (202,    22) (203,     0)]
('number', 5540)
[(200, 11745) (201, 31689) (202,     1) (203,     0)]
('number', 5541)
[(200, 36636) (201, 31689) (202,     2) (203,     0)]
('number', 5542)
[(200, 32415) (201, 31693) (202,    21) (203,     0)]
('number', 5543)
[(200, 33891) (201, 31694) (202,     6) (203,     0)]
('number', 5544)
[(200, 62057) (201, 31695) (202,     7) (203,     0)]
('number', 5545)
[(200, 53026) (201, 31696) (202,     5) (203,     0)]
('number', 5546)
[(200,    81) (201, 31711) (202,    76) (203,     0)]
('number', 5547)
[(200, 29261) (201, 31717) (202,    34) (203,     0)]
('number', 5548)
[(200, 21932) (201, 31718) (202,     5) (203,     0)]
('number', 5549)
[(200,  1878) (201, 31721) (202,    14) (203,     0)]
('number', 5550)
[(200,  1797) (201, 31722) (202,     6) (203,     0)]
('numb

('number', 5817)
[(200, 18126) (201, 32490) (202,    39) (203,     0)]
('number', 5818)
[(200,  1302) (201, 32491) (202,     4) (203,     0)]
('number', 5819)
[(200, 64130) (201, 32495) (202,    26) (203,     0)]
('number', 5820)
[(200, 52197) (201, 32496) (202,     5) (203,     0)]
('number', 5821)
[(200, 21710) (201, 32498) (202,     8) (203,     0)]
('number', 5822)
[(200, 65477) (201, 32499) (202,     9) (203,     0)]
('number', 5823)
[(200, 16750) (201, 32501) (202,     7) (203,     0)]
('number', 5824)
[(200, 49737) (201, 32503) (202,    13) (203,     0)]
('number', 5825)
[(200,  1879) (201, 32505) (202,     7) (203,     0)]
('number', 5826)
[(200, 19590) (201, 32507) (202,    12) (203,     0)]
('number', 5827)
[(200, 12200) (201, 32509) (202,    10) (203,     0)]
('number', 5828)
[(200, 24885) (201, 32510) (202,     6) (203,     0)]
('number', 5829)
[(200,  9883) (201, 32512) (202,    10) (203,     0)]
('number', 5830)
[(200, 58628) (201, 32514) (202,    14) (203,     0)]
('numb

('number', 6081)
[(200, 22848) (201, 33368) (202,    43) (203,     0)]
('number', 6082)
[(200,  5499) (201, 33371) (202,    15) (203,     0)]
('number', 6083)
[(200, 42964) (201, 33373) (202,    13) (203,     0)]
('number', 6084)
[(200, 64432) (201, 33375) (202,    13) (203,     0)]
('number', 6085)
[(200, 19235) (201, 33376) (202,     1) (203,     0)]
('number', 6086)
[(200, 34080) (201, 33379) (202,    18) (203,     0)]
('number', 6087)
[(200, 22868) (201, 33392) (202,    68) (203,     0)]
('number', 6088)
[(200, 24675) (201, 33407) (202,    81) (203,     0)]
('number', 6089)
[(200,  3833) (201, 33408) (202,     3) (203,     0)]
('number', 6090)
[(200, 55592) (201, 33410) (202,    15) (203,     0)]
('number', 6091)
[(200, 36327) (201, 33413) (202,    15) (203,     0)]
('number', 6092)
[(200, 33056) (201, 33415) (202,    10) (203,     0)]
('number', 6093)
[(200, 63781) (201, 33415) (202,     2) (203,     0)]
('number', 6094)
[(200, 49722) (201, 33419) (202,    21) (203,     0)]
('numb

[(200, 56105) (201, 34296) (202,    11) (203,     0)]
('number', 6346)
[(200, 60497) (201, 34297) (202,     5) (203,     0)]
('number', 6347)
[(200, 38533) (201, 34298) (202,     4) (203,     0)]
('number', 6348)
[(200, 51196) (201, 34301) (202,    17) (203,     0)]
('number', 6349)
[(200, 57228) (201, 34303) (202,    11) (203,     0)]
('number', 6350)
[(200, 13557) (201, 34304) (202,     2) (203,     0)]
('number', 6351)
[(200, 22236) (201, 34305) (202,     6) (203,     0)]
('number', 6352)
[(200, 38901) (201, 34306) (202,     7) (203,     0)]
('number', 6353)
[(200, 17060) (201, 34309) (202,    14) (203,     0)]
('number', 6354)
[(200, 28761) (201, 34309) (202,     1) (203,     0)]
('number', 6355)
[(200,  4710) (201, 34310) (202,     3) (203,     0)]
('number', 6356)
[(200, 33798) (201, 34315) (202,    30) (203,     0)]
('number', 6357)
[(200, 19777) (201, 34316) (202,     4) (203,     0)]
('number', 6358)
[(200, 57951) (201, 34316) (202,     3) (203,     0)]
('number', 6359)
[(200,

('number', 6616)
[(200, 17610) (201, 35124) (202,    10) (203,     0)]
('number', 6617)
[(200, 20770) (201, 35130) (202,    32) (203,     0)]
('number', 6618)
[(200,  4530) (201, 35138) (202,    41) (203,     0)]
('number', 6619)
[(200, 20606) (201, 35138) (202,     1) (203,     0)]
('number', 6620)
[(200, 50015) (201, 35139) (202,     8) (203,     0)]
('number', 6621)
[(200, 28155) (201, 35140) (202,     4) (203,     0)]
('number', 6622)
[(200,  5976) (201, 35142) (202,     8) (203,     0)]
('number', 6623)
[(200, 47967) (201, 35143) (202,     9) (203,     0)]
('number', 6624)
[(200, 63393) (201, 35144) (202,     7) (203,     0)]
('number', 6625)
[(200, 17318) (201, 35145) (202,     1) (203,     0)]
('number', 6626)
[(200, 34286) (201, 35150) (202,    29) (203,     0)]
('number', 6627)
[(200,  1387) (201, 35156) (202,    29) (203,     0)]
('number', 6628)
[(200, 10854) (201, 35158) (202,    11) (203,     0)]
('number', 6629)
[(200, 62120) (201, 35160) (202,    15) (203,     0)]
('numb

[(200,  9730) (201, 35960) (202,    22) (203,     0)]
('number', 6880)
[(200, 58333) (201, 35960) (202,     4) (203,     0)]
('number', 6881)
[(200, 16697) (201, 35963) (202,    12) (203,     0)]
('number', 6882)
[(200, 41595) (201, 35970) (202,    40) (203,     0)]
('number', 6883)
[(200,  6846) (201, 35977) (202,    34) (203,     0)]
('number', 6884)
[(200,  9244) (201, 35983) (202,    33) (203,     0)]
('number', 6885)
[(200, 54312) (201, 35988) (202,    30) (203,     0)]
('number', 6886)
[(200, 11801) (201, 35990) (202,     7) (203,     0)]
('number', 6887)
[(200, 42727) (201, 35991) (202,     8) (203,     0)]
('number', 6888)
[(200, 35854) (201, 35992) (202,     5) (203,     0)]
('number', 6889)
[(200, 17602) (201, 35996) (202,    20) (203,     0)]
('number', 6890)
[(200, 51573) (201, 36001) (202,    29) (203,     0)]
('number', 6891)
[(200, 60883) (201, 36002) (202,     6) (203,     0)]
('number', 6892)
[(200, 13628) (201, 36005) (202,    13) (203,     0)]
('number', 6893)
[(200,

('number', 7158)
[(200,  8351) (201, 36817) (202,    32) (203,     0)]
('number', 7159)
[(200, 55061) (201, 36822) (202,    31) (203,     0)]
('number', 7160)
[(200, 38241) (201, 36823) (202,     4) (203,     0)]
('number', 7161)
[(200, 62641) (201, 36823) (202,     2) (203,     0)]
('number', 7162)
[(200, 57150) (201, 36826) (202,    15) (203,     0)]
('number', 7163)
[(200, 41478) (201, 36827) (202,     4) (203,     0)]
('number', 7164)
[(200, 50915) (201, 36828) (202,     6) (203,     0)]
('number', 7165)
[(200, 34136) (201, 36829) (202,     4) (203,     0)]
('number', 7166)
[(200, 59748) (201, 36829) (202,     2) (203,     0)]
('number', 7167)
[(200, 50249) (201, 36830) (202,     5) (203,     0)]
('number', 7168)
[(200, 57222) (201, 36837) (202,    38) (203,     0)]
('number', 7169)
[(200,  2680) (201, 36847) (202,    49) (203,     0)]
('number', 7170)
[(200, 12047) (201, 36857) (202,    54) (203,     0)]
('number', 7171)
[(200, 51659) (201, 36857) (202,     3) (203,     0)]
('numb

('number', 7436)
[(200,  1910) (201, 37679) (202,     3) (203,     0)]
('number', 7437)
[(200,  4800) (201, 37681) (202,    11) (203,     0)]
('number', 7438)
[(200, 18216) (201, 37681) (202,     1) (203,     0)]
('number', 7439)
[(200, 40853) (201, 37684) (202,    17) (203,     0)]
('number', 7440)
[(200, 35929) (201, 37685) (202,     5) (203,     0)]
('number', 7441)
[(200, 30799) (201, 37688) (202,    16) (203,     0)]
('number', 7442)
[(200,   405) (201, 37691) (202,    14) (203,     0)]
('number', 7443)
[(200, 55061) (201, 37694) (202,    20) (203,     0)]
('number', 7444)
[(200, 40241) (201, 37700) (202,    31) (203,     0)]
('number', 7445)
[(200,  5443) (201, 37704) (202,    19) (203,     0)]
('number', 7446)
[(200,   836) (201, 37707) (202,    15) (203,     0)]
('number', 7447)
[(200,  6723) (201, 37708) (202,     6) (203,     0)]
('number', 7448)
[(200,   162) (201, 37710) (202,    10) (203,     0)]
('number', 7449)
[(200, 14890) (201, 37710) (202,     1) (203,     0)]
('numb

('number', 7709)
[(200, 15025) (201, 38571) (202,    11) (203,     0)]
('number', 7710)
[(200, 62794) (201, 38576) (202,    31) (203,     0)]
('number', 7711)
[(200, 38151) (201, 38577) (202,     3) (203,     0)]
('number', 7712)
[(200, 48911) (201, 38581) (202,    23) (203,     0)]
('number', 7713)
[(200, 43495) (201, 38588) (202,    36) (203,     0)]
('number', 7714)
[(200, 62538) (201, 38592) (202,    23) (203,     0)]
('number', 7715)
[(200, 18762) (201, 38594) (202,     8) (203,     0)]
('number', 7716)
[(200, 35380) (201, 38597) (202,    17) (203,     0)]
('number', 7717)
[(200, 46302) (201, 38598) (202,     6) (203,     0)]
('number', 7718)
[(200, 52529) (201, 38599) (202,     6) (203,     0)]
('number', 7719)
[(200, 38376) (201, 38600) (202,     4) (203,     0)]
('number', 7720)
[(200, 51694) (201, 38601) (202,     7) (203,     0)]
('number', 7721)
[(200, 47922) (201, 38605) (202,    21) (203,     0)]
('number', 7722)
[(200, 26535) (201, 38613) (202,    41) (203,     0)]
('numb

('number', 7963)
[(200, 40664) (201, 39492) (202,     8) (203,     0)]
('number', 7964)
[(200, 59318) (201, 39494) (202,    12) (203,     0)]
('number', 7965)
[(200, 27130) (201, 39495) (202,     3) (203,     0)]
('number', 7966)
[(200, 61891) (201, 39500) (202,    30) (203,     0)]
('number', 7967)
[(200, 35110) (201, 39508) (202,    40) (203,     0)]
('number', 7968)
[(200, 59990) (201, 39512) (202,    24) (203,     0)]
('number', 7969)
[(200, 41973) (201, 39516) (202,    19) (203,     0)]
('number', 7970)
[(200, 55710) (201, 39527) (202,    60) (203,     0)]
('number', 7971)
[(200, 58581) (201, 39528) (202,     6) (203,     0)]
('number', 7972)
[(200, 33782) (201, 39535) (202,    35) (203,     0)]
('number', 7973)
[(200, 51540) (201, 39537) (202,    12) (203,     0)]
('number', 7974)
[(200, 30676) (201, 39544) (202,    36) (203,     0)]
('number', 7975)
[(200, 64883) (201, 39544) (202,     3) (203,     0)]
('number', 7976)
[(200, 60196) (201, 39545) (202,     5) (203,     0)]
('numb

In [5]:
import AcquDetector as aqdet

# AcquDetector module
The detector module, holds all of the information on the detectors and how to convert the raw adc values into relevent detector signals.

Detector information has been converted into the JSON file format. Currently with one file for each measurement parameter (e.g. NaI Energy, NaI Time, BaF Energy, BaF SG Energy, BaF Time, Tagger Time)

The JSON format:
    
    detector   : Unique ID String (e.g. NaI or CB)
    calEq      : python equation which applies calibration to adcs
    channels   : array of channels
        channel    : channel number
        adc        : adc number relating to measurement
        raw        : empty python array raw adc value will be entered into
        value      : empty python array calibrated value will be entered into
        ref        : array number of reference channel if required by calEq
        neighbours : necessary if being interogated as a graph
        [consts]   : any values required by calibration or analysis
    references : array of required reference adcs
        adc        : adc number relating to reference channel
        raw        : empty python array raw adc value will be entered into
       
Acqu setup files can be easily converted using python scripts.

The calEq which allows the detector to self calibrate is possibly not very user friendly.

An example of the first channel of the CB for Mk1 Data:

{
  "detector": "CB", 
  "parameter": "Energy", 
  "calEq": "return [(float(y['raw'][1]-y['raw'][0]))*y['scale']+y['offset'] for y in x]"
  "channels": [
    {
      "scale": 0.0605896, 
      "value": [], 
      "raw": [], 
      "neighbours": [
        1, 
        144, 
        576, 
        145, 
        577, 
        432, 
        579, 
        2, 
        3, 
        146, 
        288
      ], 
      "adc": 3015, 
      "offset": 0.2, 
      "position": [
        3.619, 
        45.214, 
        2.629
      ], 
      "channel": 0
    }
}

Detector files are loaded in a list or file with the file locations:

In [8]:
aqdet.LoadDetectors(['aux/CrystalBallEnergy.json','aux/TAPSEnergy.json','aux/tagger.json'])
#aqdet.LoadDetectors(['aux/CrystalBallEnergy.json','aux/TAPSEnergyDom.json'])
#aqdet.LoadDetectorsFile(detFile)

aux/CrystalBallEnergy.json
aux/TAPSEnergy.json
aux/tagger.json


# Calibrating the adc values

Calling Calibrate on the raw adc array filters the array into the different detectors loaded and applies the calibration equation returning a python dictionary containing a numpy array with all of each channels parameters.

In [9]:
import ROOT
import numpy as np
from rootpy.plotting import histogram, Hist2D, Hist, Canvas

taggerChannels = aqdet.Channels['tagger']
taggerchan     = Hist(taggerChannels,0,taggerChannels,name='taggerChan')
CBHitsE = Hist2D(200,0,200,200,0,2000,name='CBHitsE')
CBChanE = Hist2D(aqdet.Channels['CB'],0,aqdet.Channels['CB'],100,0,1000,name='CBEChan')

def plotCalCB():
    data = aqdet.Calibrate(aq.adcArray)    
    CBHitsE.Fill(data['CB']['value'].size,np.sum(data['CB']['value']))
    for dat in data['CB']:
        CBChanE.Fill(dat['channel'],dat['value'])
    for dat in data['tagger']:
        taggerchan.Fill(dat['channel'])
        
    if(aq.eventNo%1000==0):
        print("Events Processed: ",aq.eventNo)
    
aq.runFunction(plotCalCB,0,5000)

ROOT.enableJSVis()
c2 = ROOT.TCanvas("my3","The Canvas Title",800,600)
taggerchan.Draw("colz")
c2.Draw()
c0 = ROOT.TCanvas("my","The Canvas Title",800,600)
CBHitsE.Draw("colz")
c0.Draw()
c1 = ROOT.TCanvas("my2","The Canvas Title",800,600)
CBChanE.Draw("colz")
c1.Draw()

/home/clas12/anaconda2/lib/python2.7/site-packages/numpy/core/records.py:513: FutureWarning:

Numpy has detected that you may be viewing or writing to an array returned by selecting multiple fields in a structured array. 

This code may break in numpy 1.15 because this will return a view instead of a copy -- see release notes for details.

AcquDetector.py:85: FutureWarning:

Numpy has detected that you (may be) writing to an array returned
by numpy.diagonal or by selecting multiple fields in a structured
array. This code will likely break in a future numpy release --
see numpy.diagonal or arrays.indexing reference docs for details.
The quick fix is to make an explicit copy (e.g., do
arr.diagonal().copy() or arr[['f0','f1']].copy()).



('Events Processed: ', 1000)
('Events Processed: ', 2000)
('Events Processed: ', 3000)
('Events Processed: ', 4000)
('Events Processed: ', 5000)


# Plotting detector graphs

The plotly python package has some very nice plotting capabilities. The detectors verteces and neighbour edges can easily be plotted in 3D to examine their structures. Here we see my TAPS file has some problems!

In [ ]:
import plotly.offline as py
import plotly.graph_objs as go
py.init_notebook_mode()

# CB plot
aqdet.detgraphs['CB'].vs['label'] = aqdet.detgraphs['CB'].vs['channel']
FullCBnodes, FullCBedges     = aqdet.GraphPlot(aqdet.detgraphs['CB'])

# TAPS plot
aqdet.detgraphs['TAPS'].vs['label'] = aqdet.detgraphs['TAPS'].vs['channel']
FullTAPSnodes, FullTAPSedges = aqdet.GraphPlot(aqdet.detgraphs['TAPS'])

layout = go.Layout(
    title = 'title',
    width=1000,
    height=1000
)

figCB   = go.Figure(data=[FullCBnodes,FullCBedges],layout=layout)

figTAPS = go.Figure(data=[FullTAPSnodes,FullTAPSedges],layout=layout)

py.iplot(figCB)
py.iplot(figTAPS)

# Drawing detector hits

The way the detector and data is output makes it easy to draw the real data (as a sub-graph of the detector) ontop of the defined detector geometries. Here **CalGraph** is called instead of **calibrate** returning a graph for each event rather than an array. The graph package **igraph** has clustering algorithms but here each non-joined set is shown in a different colour.

In [ ]:
import math

def DrawDetectorHits():
    print('number',aq.eventNo)
    #print(A2G4.signals)
    
    data = aqdet.CalGraph(aq.adcArray)
            
    plots  = [aqdet.GraphPlot(plot,1)[0] for plot in data['CB'].decompose()]
    plots += [aqdet.GraphPlot(plot,1)[0] for plot in data['TAPS'].decompose()]
    plots += [aqdet.GraphPlot(plot,1)[1] for plot in data['CB'].decompose()]
    plots += [aqdet.GraphPlot(plot,1)[1] for plot in data['TAPS'].decompose()]
    
    fig = go.Figure(data=[FullCBedges,FullTAPSedges,FullCBnodes,FullTAPSnodes]+plots,layout=layout)

    py.iplot(fig)
    

    
    
    

In [ ]:
aqdet.LoadDetectors(['aux/CrystalBallEnergy.json','aux/TAPSEnergy.json'])
maxEvnts = 28
aq.runFunction(DrawDetectorHits,25,maxEvnts)

# A2G4 module

Reads in the MC file values requested. As above we can run any function on the event data, now with easy access to the geant4 input parameters too. When MCCalGraph is called a list of input particles to use can also be passed in order to mask any signals. (Be warned my version of the simulation keeps track of what exactly energy belonged to what primary particle).

**Simulation output is a photon and proton - The colour of the points is now given by which input particle deposited the energy**

In [ ]:
import A2G4

mcFile = '/w/work1/home/simong/Simulation/G4Out/GP_GP_20994.root'
A2G4.openFile(mcFile)

In [ ]:
def DrawDetectorHitsMC():
    print('number',A2G4.eventNo)
    data = aqdet.MCCalGraph(A2G4.signals,[1])
    
    plots  = [plot for plot in aqdet.GraphPlot(data['CB'],1)]
    plots += [plot for plot in aqdet.GraphPlot(data['TAPS'],1)]
    
    data  = aqdet.MCCalGraph(A2G4.signals,[2])
    
    plots += [plot for plot in aqdet.GraphPlot(data['CB'],1)]
    plots += [plot for plot in aqdet.GraphPlot(data['TAPS'],1)]    
    plots += [aqdet.InputPlot(A2G4.inpart)[0]]
    plots += [aqdet.InputPlot(A2G4.inpart)[1]]

    print(A2G4.inpart['plab']*1000)
    
    fig = go.Figure(data=[FullCBedges,FullTAPSedges,FullCBnodes,FullTAPSnodes]+plots,layout=layout)

    py.iplot(fig)

In [ ]:
maxEvnts = 12
A2G4.runFunction(DrawDetectorHitsMC,10,maxEvnts)